### PR TITLE
Fix mypy errors around BalanceProof

### PR DESCRIPTION
**Pull request by pirapira**
The biggest change was to allow None as the balance_proof field of
a ContractSendChannelClose event.

This is a part of https://github.com/raiden-network/raiden/issues/3040

### CODE REVIEW

File: raiden/transfer/channel.py [hunk 1]
```diff
@@ -1310,6 +1312,9 @@ def events_for_close(
             None,
         )
 
+        # silence mypy: partner's balance proofs should be signed
+        assert not isinstance(channel_state.partner_state.balance_proof, BalanceProofUnsignedState)
```
Comment (LefterisJP) [inline raiden/transfer/channel.py]: Hmm I am not sure about this one.
First if you wanted an assertion here, you should have a positive and not negative assertion. So `isinstance(channel_state.partner_state.balance_proof, BalanceProofSignedState)`.

But secondly when a `NettingChannelState` and a `NettingChannelEndState` is [initialized](https://github.com/raiden-network/raiden/blob/f5d48172931ce60c820253e81708acc2a2f49941/raiden/blockchain/state.py#L8) the initial `balance_proof` is `None`. The type of the balance proof considers `None` a valid value.

And it definitely is a valid value since the partner may have sent us no transfers. In which case when we handle the Channel close (see below) we sent empty hashes.

https://github.com/raiden-network/raiden/blob/f5d48172931ce60c820253e81708acc2a2f49941/raiden/raiden_event_handler.py#L241-L258

This probably also did not hit in any of the tests since the negative assertions holds for any type. e.g. 

```
True == not isinstance(str, BalanceProofUnsignedState)
True == not isinstance(None, BalanceProofUnsignedState)
```
e.t.c.

P.S.: That means that `ContractSendChannelClose` constructor also need the `balance_proof` argument to be `typing.Optional[BalanceProofSignedState]` and not `BalanceProofSignedState`.

File: raiden/transfer/channel.py [hunk 2]
```diff
@@ -1334,8 +1339,11 @@ def create_sendexpiredlock(
     nonce = get_next_nonce(sender_end_state)
     locked_amount = get_amount_locked(sender_end_state)
     balance_proof = sender_end_state.balance_proof
-    updated_locked_amount = locked_amount - locked_lock.amount
-    transferred_amount = balance_proof.transferred_amount
+    updated_locked_amount: typing.TokenAmount = locked_amount - locked_lock.amount
+    if balance_proof:
```
Comment (LefterisJP) [inline raiden/transfer/channel.py]: Hmm here I believe the `sender_end_state` should always contain a balance_proof since it's `our_state` and this is only generated for a lock expiration so there should be a balance proof. So I would replace this with an assertion.

```python
assert balance_proof, 'The sender end state should always have a balance proof at this point'
```

File: raiden/transfer/channel.py [hunk 3]
```diff
@@ -1705,6 +1713,10 @@ def handle_channel_closed(
         )
         if call_update:
             expiration = state_change.block_number + channel_state.settle_timeout
+            # silence mypy: checked above already
+            assert balance_proof is not None
```
Comment (LefterisJP) [inline raiden/transfer/channel.py]: Hmmm this is a double assertion. Right above inside `call_update` we are checking that `balance_proof` is not None. Please find a way to not do this twice.

File: raiden/transfer/channel.py [hunk 4]
```diff
@@ -1705,6 +1713,10 @@ def handle_channel_closed(
...
+            # silence mypy: partner's balance proof is always signed
+            assert not isinstance(balance_proof, BalanceProofUnsignedState)
```
Comment (LefterisJP) [inline raiden/transfer/channel.py]: As per my original comment please have a positive assertion (That this is a `BalaneProofSignedState`).

File: raiden/transfer/mediated_transfer/events.py [hunk 5]
```diff
@@ -305,10 +305,10 @@ def from_dict(cls, data: typing.Dict[str, typing.Any]) -> 'SendBalanceProof':
         restored = cls(
             recipient=to_canonical_address(data['recipient']),
             channel_identifier=typing.ChannelID(int(data['channel_identifier'])),
-            message_identifier=int(data['message_identifier']),
-            payment_identifier=int(data['payment_identifier']),
+            message_identifier=typing.MessageID(int(data['message_identifier'])),
```
Comment (LefterisJP) [inline raiden/transfer/mediated_transfer/events.py]: Both changes to ID type should be fine since in the runtime they are all ints anyway.
So I guess doing something like: `typing.MessageID(data['message_identifier'])` would also work?

File: raiden/transfer/state.py [hunk 6]
```diff
@@ -805,9 +805,9 @@ def to_dict(self) -> typing.Dict[str, typing.Any]:
     def from_dict(cls, data: typing.Dict[str, typing.Any]) -> 'BalanceProofUnsignedState':
         restored = cls(
             nonce=data['nonce'],
-            transferred_amount=int(data['transferred_amount']),
-            locked_amount=int(data['locked_amount']),
-            locksroot=serialization.deserialize_bytes(data['locksroot']),
+            transferred_amount=typing.TokenAmount(int(data['transferred_amount'])),
```
Comment (LefterisJP) [inline raiden/transfer/state.py]: perhaps try the same thing I said above for ints here and below?

File: raiden/transfer/channel.py (hunk 1 above)
Comment (pirapira) [reply raiden/transfer/channel.py]: That's why I wanted somebody's eye on this one.

I'll change the assertion to positive, but it has to allow `None` in addition to `BalanceProofSignedState`.

> P.S.: That means that ContractSendChannelClose constructor also need the balance_proof argument to be typing.Optional[BalanceProofSignedState]

Yes, mypy forced me to make that change (already in this PR)
https://github.com/raiden-network/raiden/pull/3214/files#diff-dbf5b17f182eaefb153df563cd0418deR27

File: raiden/transfer/channel.py (hunk 2 above)
Comment (pirapira) [reply raiden/transfer/channel.py]: I see. There has to be a lock before it expires.

File: raiden/transfer/channel.py (hunk 1 above)
Comment (pirapira) [reply raiden/transfer/channel.py]: Done.  Is this what you wanted? https://github.com/raiden-network/raiden/pull/3214/files#diff-fe7772da6084a3493b31e88fbf7f6d55R1317 @LefterisJP

File: raiden/transfer/channel.py (hunk 2 above)
Comment (pirapira) [reply raiden/transfer/channel.py]: Done https://github.com/raiden-network/raiden/pull/3214/files#diff-fe7772da6084a3493b31e88fbf7f6d55R1345

Comment (LefterisJP) [reply raiden/transfer/channel.py]: @pirapira Since it's an assertion can you change `is` to `should be`? This is the message you will see if the assertion fails.

File: raiden/transfer/channel.py (hunk 1 above)
Comment (LefterisJP) [reply raiden/transfer/channel.py]: yes thank you.

File: raiden/transfer/channel.py (hunk 2 above)
Comment (pirapira) [reply raiden/transfer/channel.py]: You're absolutely right.

File: raiden/transfer/channel.py [hunk 7]
```diff
@@ -1310,11 +1312,15 @@ def events_for_close(
             None,
         )
 
+        balance_proof = channel_state.partner_state.balance_proof
+        # silence mypy: partner's balance proofs should be signed
```
Comment (hackaugusto) [inline raiden/transfer/channel.py]: IMO the correct fix is to split the `NettingChannelEndState` class into two, one for our state and another for the partner's state. If we do split it, once we have annotations for most of the code base we can then remove the runtime type assertions. (Note: As far as I recall we only have not done it yet to safe time)

Comment (pirapira) [reply raiden/transfer/channel.py]: Created an issue for that https://github.com/raiden-network/raiden/issues/3222
